core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -286,15 +286,18 @@ class CassandraJournal(cfg: Config)
         result.map(_ => extractTagWrites(serialized))
     }
 
-    tws.onComplete { result =>
-      self ! WriteFinished(pid, p.future)
-      p.success(Done)
-      // notify TagWriters when write was successful
-      result.foreach(bulkTagWrite => tagWrites.foreach(_ ! bulkTagWrite))
+    bulkTagWrite.onComplete { _ =>
+      self ! WriteFinished(pid, writeInProgressForPersistentId.future)
+      writeInProgressForPersistentId.success(Done)
     }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
 
     //Nil == all good
-    tws.map(_ => Nil)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+    bulkTagWrite.map(bulkTagWrite => {
+      // notify TagWriters when write was successful before completing the future meaning
+      // that we can get another seq of AtomicWrites for the same persistent actor
+      tagWrites.foreach(_ ! bulkTagWrite)
+      Nil
+    })(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
   }
```

Comment:
I don't understand this. Isn't `bulkTagWrite.onComplete` and `bulkTagWrite.map` competing (undefined order)?

also confusing to use the same name twice in `bulkTagWrite.map(bulkTagWrite =>`

Reply:
It doesn't matter when the `WriteFinished` is sent (it is in the `onComplete`) but it does matter that the message is sent to the tag writer before the Future that is returned completes.

The message to the tag writer previously was in an onComplete meaning that it was not guaranteed to happen before another call to `asyncWriteMessages` for the same persistenceId.

I'll just move everything to the map to make this simpler as there's nothing wrong with sending the `WriteFinished` before we complete the future returned to the `AsyncWriteJournal`.

Does that make sense @patriknw ?

Reply:
Also will rename the inner bulkTagWrite

Reply:
ok, also good to complete the `writeInProgressForPersistentId` afterwards then, because otherwise another asyncWriteMessages could theoretically be started before in case of a restart (possibly, maybe?)